### PR TITLE
[scheduler-premium] Add the dependencies data model

### LINKED ISSUE #22854: [scheduler] Dependencies: data model
Part of #22853 (event-scheduling umbrella) — FS behavior.

Define the data structure for event dependencies. The shape only ships what is implemented: the `type` union starts with `'FinishToStart'`, and the other PDM types plus `lag`/`lagUnit` are added together with their implementations — widening a union and adding optional fields are additive changes, so nothing needs to be reserved up front.

A dependency is **not** a field on the event — it's a separate relationship (a graph edge) between two events, referencing them by `id`. One event can take part in many dependencies.

```ts
type SchedulerDependencyId = string | number; // same shape as SchedulerEventId

interface SchedulerDependency {
  id: SchedulerDependencyId;
  source: SchedulerEventId; // predecessor event id (Bryntum: `from`)
  target: SchedulerEventId; // successor event id   (Bryntum: `to`)
  type: 'FinishToStart'; // union widened with 'StartToStart' | 'FinishToFinish' | 'StartToFinish' when implemented
}
```

**Lag (design decided now, fields added when implemented):** the fields are deliberately NOT reserved in the type — an unimplemented field in the public shape is a silent no-op promise (`lag: 2` doing nothing invites bug reports), and adding optional fields later is non-breaking anyway. What IS decided now is the design for when it lands: `lag` alone would under-specify it — every engine pairs the amount with a unit (Bryntum: `lag` + `lagUnit`, default `'day'`; Syncfusion: `"+3d"/"+3h"/"+3m"`; MS Project: working-time days vs elapsed `6ed`). Since the scheduler has no working-calendar concept, lag will mean **elapsed (calendar) time**, applied through the date adapter (DST-safe), with the unit made explicit via `lagUnit`.

## Where everything lives (three levels)

| What | Where |
| --- | --- |
| **Data** — the dependencies collection | reactive store state → `EventTimelinePremiumState` |
| **Hook / delegation** on event create/update/delete | `SchedulerStore` (interface, recurring-events pattern) |
| **Logic / engine** — cascade, mutations | the scheduling plugin (premium-internals) |

### Data — in the reactive store state
Dependencies are timeline-premium-only, and components render the arrows through **selectors** that subscribe to the reactive state — so the collection must live in state, not as a private plugin field. Store it in `EventTimelinePremiumState` (which `extends SchedulerState`), following the events naming convention (`eventModelList` / `eventModelLookup`) but with only the two fields dependencies actually need:

```ts
interface EventTimelinePremiumState extends SchedulerState {
  // ...
  dependencyModelList: readonly SchedulerDependency[];                      // prop mirror / iteration
  dependencyModelLookup: Map<SchedulerDependencyId, SchedulerDependency>;   // lookup by id
}
```

No separate id list (unlike events): dependencies have no semantic order and no model/processed split, and a `Map` preserves insertion order, so `lookup.keys()` already yields ordered ids if a selector ever needs them.

The graph indexes the engine needs — adjacency by `source` (cascade) and by `target` (cascade-delete, cycle guard) — are derivable, so they live in **memoized selectors** over `dependencyModelList`, not in state.

**Controlled prop, like events.** Events are a controlled prop: state mirrors `parameters.events`, and mutations don't write state — they emit the new array via `onEventsChange`. Dependencies follow the same pattern: a `dependencies` parameter synced into the slice above + an `onDependenciesChange` callback emitted by the mutation paths.

### Hook — in `SchedulerStore` (community base)
A `SchedulerSchedulingPluginInterface` + an optional injected `schedulingPlugin` ref, with delegation hooks in the event create/update/delete path (the cascade must trigger when an event moves, and event updates flow through the base store — same as `recurringEventsPlugin`). Premium-agnostic.

Note the seam: the delegation must run **inside `updateEvents`, before `onEventsChange`**, so cascaded changes merge into the same atomic batch. The `eventsUpdated` published event (the lazy-loading seam) fires in a `queueMicrotask` *after* `onEventsChange` and is not suitable for this (details in #22857).

### Logic — in the plugin (premium-internals)
The actual implementation: reads/mutates the `dependencies` state slice via `store.set(...)`, plus the cascade engine (and later constraint reconciliation). Holds only transient data privately, exactly like `SchedulerLazyLoadingPlugin` keeps its cache/queue private while writing canonical data into store state. Injected only for premium → no premium code in the community package. This shared home is what lets constraints (used by both premium products, which don't share a premium store base) reuse it next quarter. See Architecture in #22853.

## Referential integrity

Dependencies reference events by id, so the model must define what happens when a reference breaks:

- **Deleting an event** also removes every dependency it takes part in (as `source` or `target`), in the same atomic update — no dangling edges.
- **Data from props** referencing an unknown event id: the dependency is kept in state but ignored by rendering and the engine. Dev-mode warning only when there's no `dataSource` — with lazy loading a missing event is expected (it may simply not be fetched yet), so a warning there would be noise.
- **Lazy loading:** the engine never fetches missing events — auto-scheduling only operates on loaded events. Documented limitation, not something the data model works around (behavior details in #22857).

## Recurring events — dependencies are disallowed in v1

Dependencies-by-event-id is unsound for recurring series:

- All occurrences of a series share the base event `id`; occurrences are computed (keyed `${id}::${dateKey}`), never materialized, so a specific occurrence isn't addressable.
- Scoped edits mint new ids ("only this" detaches a standalone event with a fresh id; "this and following" creates `${id}::${dateKey}`) — a dependency pointing at the original would silently break or point at the wrong thing.
- FS against a series has no defined semantics (which occurrence's end?).

**No competitor solves this.** Bryntum Scheduler Pro has both features but never combines them — their official recurrence demo explicitly sets `features: { dependencies: false }`, and there is no documented behavior for a dependency on a recurring event. Bryntum Gantt rejects recurrence outright ("one task equals one row"). DHTMLX and Syncfusion keep the two features in separate products (Scheduler = recurrence, Gantt = dependencies). The only products combining them (MS Project, Wrike) escape the problem by materializing each occurrence as a first-class task with its own id — and even there, linking "the series" is degenerate (MS Project resolves it to the last occurrence and doesn't draw the line).

**v1 policy:** `addDependency` rejects events with an `rrule` (with a reason, like the cycle guard), and dependencies from props referencing a recurring event are ignored + dev-mode warning.

**Open question:** should the timeline expose a prop to disable recurrence entirely, making the "Gantt mode" trade-off explicit? Prior art: Bryntum's recurrence is opt-in (`enableRecurringEvents`, default `false`); Bryntum Gantt simply doesn't support recurrence.

## Exposure — internal-only until the slice is complete

The sub-issue PRs merge incrementally while the scheduler is in beta, so they land with no public API surface — a data model without rendering or interaction would show up in the API docs as dead API, and beta users adopting it mid-slice would hit churn for nothing. No re-exports from `@mui/x-scheduler-premium`, no `buildPublicAPI` additions, and the `dependencies` / `onDependenciesChange` parameters live on an internal store-parameters type instead of `EventTimelinePremiumParameters` (which feeds the public component props, propTypes and API docs). A final PR flips the feature public deliberately once FS works end-to-end: move the two props to `EventTimelinePremiumParameters`, extend the parameters extractor, re-export the types, regenerate proptypes/docs. `addDependency`/`deleteDependency` stay internal store methods consumed by the interaction UI — the event CRUD methods (`createEvent`, `deleteEvent`, …) are not exposed through `apiRef` either, and exposing imperative CRUD would be a separate decision made for events and dependencies together.

## Scope
- Define the `SchedulerDependency` shape referencing events by id; the `type` union starts with `FinishToStart` only (see the lag note above for why nothing is reserved).
- Add the dependencies slice (model + lookup) to `EventTimelinePremiumState`, fed by a controlled `dependencies` parameter + `onDependenciesChange`; adjacency indexes as memoized selectors.
- Add the `SchedulerSchedulingPluginInterface` + injection hook in `SchedulerStore`.
- Manage the slice from the scheduling plugin; expose read/write as internal store methods (`addDependency` / `deleteDependency`) for the interaction UI, mirroring the event CRUD methods — see the exposure note above.
- Referential integrity: cascade-delete on event deletion; ignore + warn on unknown ids.
- Reject dependencies on recurring events (API + props ingestion).
- Keep the shape additive so SS/FF/SF (union widening) and `lag`/`lagUnit` (optional fields) can be added later with their implementations.

## Out of scope
- Rendering, interaction, and the auto-scheduling engine logic (separate sub-issues).

### PR DESCRIPTION
**Pull request by rita-codes**
Closes #22854. Part of the FS dependencies vertical slice (#22853).

This PR implements the dependencies data model for the timeline. Per the exposure policy documented in #22854, everything is **internal-only**: no public props, no public API methods, no type re-exports from `@mui/x-scheduler-premium`, and `pnpm proptypes` / `pnpm docs:api` produce an empty diff. The feature flips public in a deliberate final PR once FS works end-to-end.

### What's included

- `SchedulerDependency` model (`source`/`target`/`type`) in `@mui/x-scheduler-internals-premium/models`. The shape only ships what is implemented: the `type` union starts with `'FinishToStart'`, and the other PDM types plus `lag`/`lagUnit` land together with their implementations as additive changes (see the updated notes in #22854).
- Controlled `dependencies` state slice (`dependencyModelList` + `dependencyModelLookup`) on `EventTimelinePremiumState`, fed by the internal `EventTimelinePremiumStoreParameters` type + `onDependenciesChange` — mutations never write state, mirroring the events pattern.
- Memoized selectors: `activeModelList` (filters unknown/recurring references and deduplicates by id, last-wins) and adjacency (`activeModelListBySource`/`ByTarget`) — the contract that #22855 (arrows) and #22857 (engine) will consume.
- `SchedulerSchedulingPluginInterface` seam in the community `SchedulerStore`: delegation inside `updateEvents`, before `onEventsChange`, so plugin reactions land in the same update. Community stores keep it `null` — zero behavior change.
- `SchedulerSchedulingPlugin` (premium-internals): `addDependency` (rejects `recurringEvent`/`unknownEvent` with a reason, per the v1 policy) and `deleteDependency` as internal store methods (not exposed through `apiRef`, matching the event CRUD methods), cascade-delete of dependencies when events are deleted, and dev-mode warnings (silenced for unknown ids when a `dataSource` is present).

### Notes for reviewers

- Dependencies on recurring events are rejected/ignored per the v1 policy in #22854 (no competitor combines both features).
- Known v1 limitation (documented in `handleEventsUpdate`'s JSDoc): with a `dataSource`, the dependency cascade emits before the async persistence of the event deletion, so a persist failure leaves no rollback. Design input for #22857.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by noraleonte** [comment]
# PR #23117 Review — [scheduler] Add the dependencies data model

**Verdict: solid, well-tested foundational PR.** All 495 package tests pass, TypeScript, ESLint, and Prettier are clean. It adds a controlled `dependencies` data model (types, state slice, selectors, and a `SchedulerSchedulingPlugin`) without any UI. No high-severity bugs found. Findings are minor and mostly about polish/coverage.

## Scope note

This PR is **pure data-model/store/plugin/selectors** — no components, no DOM elements, no CSS. So the review dimensions about **class names on DOM elements, CSS specificity, and `styleOverrides`/theme customization are N/A** here (nothing renders). File structure, naming, and the plugin/interface seam all follow existing conventions (`SchedulerRecurringEventsPlugin`, `SchedulerLazyLoadingPlugin`, `schedulerEventSelectors`).

## What's genuinely good

- **State slice naming** (`dependencyModelList` / `dependencyModelLookup`) mirrors `eventModelList` / `eventModelLookup`. Selector names (`modelList`, `modelLookup`, `model`) mirror `schedulerEventSelectors`. Consistent.
- **WeakMap-cached `buildDependenciesState`** correctly preserves referential stability across the per-render `updateStateFromParameters` — verified by test and by the memoized `activeModelList` selector.
- **The plugin seam** (`schedulingPlugin` protected field in base store + `SchedulerSchedulingPluginInterface`) is the right minimal contract; `handleEventsUpdate` firing inside `updateEvents` *before* `onEventsChange` is correct and tested (`calledBefore`).
- **Dedup semantics** (`new Map(...)` → last-wins, insertion order preserved) are correct and the comment accurately describes it.
- Error/warning messages follow the `MUI X Scheduler:` convention with what/why/how.

---

## Findings (prioritized)

### Low — Recurring-event dev warning is missed when the event loads *after* the dependency

`SchedulerSchedulingPlugin.ts:259-264` registers the re-validation effect **only** on `state.dependencyModelList`:

```ts
store.registerStoreEffect(
  (state) => state.dependencyModelList,
  () => this.warnOnInvalidDependencies(),
);
```

With a `dataSource`, a dependency present in props references an event that is initially `unknownEvent` (warning suppressed). When that event later lazy-loads as recurring, `processedEventLookup` changes but `dependencyModelList` does **not**, so the effect never re-runs and the `recurringEvent` warning is silently lost. The test at `dependencies.EventTimelinePremiumStore.test.ts:1019` sidesteps this by *adding* the dependency after load (which does mutate `dependencyModelList`). Dev-only impact; arguably an acceptable tradeoff to avoid re-validating on every event mutation — worth a one-line comment acknowledging it, or subscribing to `processedEventLookup` too if the completeness matters.

### Low — `modelList` and `modelLookup` selectors are exported but untested (and unused in this PR)

`eventTimelinePremiumDependencySelectors.ts:186-187`. They're trivial pass-throughs, but every other selector in the file has a test. Either add two trivial assertions or accept them as intentional API surface for the follow-up rendering PR. (`model`, `activeModelList`, `activeModelListBySource`, `activeModelListByTarget` *are* covered.)

### Low — `addDependency`/`deleteDependency` are public on both the plugin and the store

`SchedulerSchedulingPlugin.ts:304,323` are `public`, and `EventTimelinePremiumStore.ts:239-249` re-wrap them as public store methods. Two public entry points for the same operation. By contrast, `lazyLoading` methods aren't re-wrapped on the store. Consider making the plugin methods the internal ones (the store method is the public API), for a single obvious call site.

### Nit — minor coverage/consistency gaps

- `activeModelList` filtering is only exercised with a **target**-side recurring event and source/target-side unknown events; a **source**-side *recurring* dependency isn't explicitly tested (the code is symmetric, so low risk).
- `deleteDependency` is tested with a single dependency; removing one of several isn't asserted.
- `dependency-utils.ts:394` uses `Set<SchedulerDependency['id']>` while the rest of the file uses the `SchedulerDependencyId` alias — trivial consistency nit.

### Behavioral note (not a defect) — consecutive `addDependency` calls before re-render

Because `dependencies` is purely controlled, two `addDependency` calls in the same tick both read the original list, so the second emission drops the first (documented in the test at `dependencies.EventTimelinePremiumStore.test.ts:823`). This is consistent with the library's controlled model and is intentional — flagging only so it's a conscious decision.

---

## Suggested fix plan (all optional / low-effort)

1. Add a short comment at the effect registration in `SchedulerSchedulingPlugin.ts` noting the recurring-after-load warning gap (or widen the effect selector if you want full coverage).
2. Add two one-line selector tests for `modelList` / `modelLookup`, or drop them until the rendering PR needs them.
3. Decide on a single public entry point for `addDependency`/`deleteDependency` (make plugin methods internal).
4. Optionally: one test for source-side recurring rejection in `activeModelList`, one for multi-dependency `deleteDependency`, and swap `SchedulerDependency['id']` → `SchedulerDependencyId`.

Nothing here blocks merge.

**Comment by rita-codes** [comment]
Thanks for the thorough pass! Applied in `da99b9adc2`:

- **Warning gap:** added a second dev-only effect watching `processedEventLookup`, so events-side invalidation (including your lazy-loads-as-recurring case) re-warns now — plus a test.
- **`modelList`/`modelLookup`:** tests added; keeping them exported since #22855/#22856 consume them next.
- **Single entry point:** agreed — the store methods are the intended call site, and we went one step further: they stay off `apiRef` entirely, mirroring the event CRUD methods (exposure note in #22854 updated). The plugin methods now carry a "call it through the store" JSDoc pointer.
- **Nits:** source-side recurring fixture, multi-dependency `deleteDependency` test, and the `SchedulerDependencyId` alias — all done.
- **Same-tick `addDependency`:** yes, conscious decision — it's the same controlled-prop semantics events have; the test documents it.

**Review by flaviendelangle — COMMENTED** [review]
# PR review

Solid, well-tested internal-only feature — the plugin wiring, controlled emit-not-write pattern, cascade ordering (`onDependenciesChange` before `onEventsChange`), recurring detection, dedup/last-wins, and dev-warning suppression all check out. Nothing merge-blocking. The one substantive gap is that `addDependency` validates each endpoint exists and is non-recurring but performs no *semantic* validation, so it admits self-dependencies and exact-duplicate edges. Remaining findings are test gaps, small dedups, and stale doc framing.

## Bugs (2)

### 1. 🟠 `addDependency` accepts a self-dependency (`source === target`)

**Location:** `packages/x-scheduler-internals-premium/src/internals/plugins/SchedulerSchedulingPlugin.ts:336`

```ts
for (const eventId of [properties.source, properties.target]) {
  const status = classifyDependencyEvent(processedEventLookup, eventId);
  if (status !== 'ok') {
    return { status: 'rejected', reason: status, eventId };
  }
}
const dependency: SchedulerDependency = { ...properties, id: generateId('dependency') };
```

Each endpoint is classified independently; there is no `source !== target` check. A `FinishToStart` edge from an event to itself is an impossible constraint (the event must finish before it starts), and there is no cycle detection anywhere in this PR to catch it later.

**Failure scenario:** `store.addDependency({ source: 'event-a', target: 'event-a', type: 'FinishToStart' })` returns `{ status: 'added' }`, and because `[source, target].every(... === 'ok')` is true, the self-loop also passes the `activeModelList` filter and flows to rendering and the future scheduling engine (#22857) as a cyclic constraint.

**Fix:** Reject when `properties.source === properties.target` (e.g. a `'selfReference'` rejection reason, or reuse an existing reason), and add a test pinning the behavior.

### 2. 🟡 `addDependency` accepts exact-duplicate edges

**Location:** `packages/x-scheduler-internals-premium/src/internals/plugins/SchedulerSchedulingPlugin.ts:347`

No check for an existing dependency with the same `source`/`target`/`type`. Calling `addDependency({a→b})` twice yields two dependencies differing only by generated id; both survive into `activeModelList`, and `activeModelListBySource.get('a')` returns two identical edges (a double-drawn arrow once #22855 lands).

**Failure scenario:** UI wired to `addDependency` lets a user connect a→b twice; the timeline renders two overlapping arrows and the engine sees a redundant constraint.

**Fix:** Product decision — either dedup against existing edges and reject, or add an explicit test asserting duplicates are intentionally allowed.

ℹ️ Note (not a regression): two `addDependency` calls before the parent re-renders both read the same original `dependencyModelList`, so the second emission overwrites the first and the first add is lost. This is standard controlled-store behavior and is deliberately documented by the test at `dependencies.EventTimelinePremiumStore.test.ts:834`; flagging only for awareness.

## Tests (3)

### 1. 🟡 Self-dependency and duplicate-edge adds are untested

**Location:** `packages/x-scheduler-internals-premium/src/use-event-timeline-premium/tests/dependencies.EventTimelinePremiumStore.test.ts:807`

No case exercises `addDependency({source:'event-a', target:'event-a'})` or a repeated identical add. These are exactly the unguarded inputs from Bugs 1–2, so current behavior is unpinned either way.

**Failure scenario:** A future refactor adds or removes self/duplicate validation and no test notices.

### 2. 🟡 Cascade delete of an event referenced by multiple dependencies is untested

**Location:** `dependencies.EventTimelinePremiumStore.test.ts:973`

The referential-integrity test deletes `event-a`, referenced by a single dependency. `handleEventsUpdate`'s `filter` dropping several entries in one deletion pass (`SchedulerSchedulingPlugin.ts:323`) is never verified, nor the case where the deleted event is both `source` and `target` of one dependency.

**Failure scenario:** A filter regression that drops only the first matching dependency would pass all current tests.

### 3. 🟡 `activeModelListByTarget` memo stability untested; the "same instance" test is weak

**Location:** `packages/x-scheduler-internals-premium/src/event-timeline-premium-selectors/eventTimelinePremiumDependencySelectors.test.ts:125`

Only `activeModelListBySource` has a referential-stability test, and it calls the selector twice on the *same* state object — which trivially hits the memo without proving invalidation. `activeModelListByTarget` is a separate memoized selector with only a value assertion.

**Failure scenario:** A broken memo dependency on `activeModelListByTarget` (or lost stability across unrelated state updates) ships unnoticed.

**Fix:** Add a target-side stability test and thread two `updateStateFromParameters` calls (unchanged deps → same ref; changed deps → new ref).

## Simplifications (1)

### 1. 🟡 The `[source, target]` classify loop and the emit-if-changed block are each duplicated

**Location:** `packages/x-scheduler-internals-premium/src/internals/plugins/SchedulerSchedulingPlugin.ts:340`, `:358`, and `eventTimelinePremiumDependencySelectors.ts:202`

Three sites iterate `[dependency.source, dependency.target]` through `classifyDependencyEvent` (add, warn, selector filter), and `handleEventsUpdate` + `deleteDependency` both repeat `remaining = current.filter(...); if (remaining.length !== current.length) this.updateDependencies(remaining)`.

**Cost:** The "what makes a dependency invalid" rule lives in three places, so Bug 1's self-check would have to be added to each. A `getDependencyRejection(processedEventLookup, endpoints)` helper (returns `{reason, eventId} | null`) and a `replaceDependenciesIfChanged(remaining)` method would centralize both.

## Docs (2)

### 1. 🟡 `source`/`target` documented as "predecessor"/"successor" won't survive new dependency types

**Location:** `packages/x-scheduler-internals-premium/src/models/dependency.ts:491`

```ts
/** The id of the predecessor event. */
source: SchedulerEventId;
/** The id of the successor event. */
target: SchedulerEventId;
```

The temporal predecessor/successor framing only holds for `FinishToStart`. Once `StartToStart`/`FinishToFinish`/`StartToFinish` widen `SchedulerDependencyType` (as the adjacent comment on line 480 promises), `source` is no longer necessarily the temporal predecessor. Describe the graph direction (source/target, driving/driven), not the temporal role.

### 2. ℹ️ `classifyDependencyEvent`'s return union is structurally coupled to `SchedulerDependencyRejectionReason` without being typed as such

**Location:** `packages/x-scheduler-internals-premium/src/internals/utils/dependency-utils.ts:453`

`addDependency` does `reason: status` relying on TS narrowing the non-`'ok'` members to exactly `SchedulerDependencyRejectionReason`. Correct today, but adding a future classification that is *not* a rejection reason (e.g. `'archivedEvent'`) would silently leak it as a `reason`. Typing the classifier as `SchedulerDependencyRejectionReason | 'ok'` (defined together) would enforce the relationship. Optional.

## Verdict

**Approve after nits** — no blocking issues; the self-dependency gap (Bug 1) and the paired test gap are the two worth addressing before merge, the rest are optional cleanups.

---

🤖 Review generated with Claude Code

**Comment by rita-codes** [comment]
Thanks! Applied in `b7428ec5b8`:

- **Self-dependency:** deferred to #22858 — a self-loop is the degenerate case of a cycle, and the guard's reachability check catches it as its base case; rejecting it here would mint a rejection reason the guard would immediately supersede. Pinned today's accept behavior with a test pointing at #22858, and left a note on that issue so it's explicit in its scope.
- **Duplicate edges:** decided to reject — `addDependency` now returns `{ status: 'rejected', reason: 'duplicateDependency', dependencyId }` pointing at the existing edge (identity = `source` + `target` while the type union has a single member). Props keep tolerating consumer data, same as recurring references. Test added.
- **Tests:** multi-dependency cascade delete, and `activeModelListByTarget` referential stability with real invalidation across `updateStateFromParameters` calls.
- **Docs:** `source`/`target` reworded to graph direction with the FS mapping noted once on the interface; `classifyDependencyEvent` is now typed as `SchedulerDependencyRejectionReason | 'ok'`, making the coupling structural.
- **Simplification:** took the emit-if-changed half (shared by the cascade and `deleteDependency`). Passed on the shared rejection helper — the three classify sites deliberately differ in reaction, and the new self/duplicate rules are API-only: they don't apply to props ingestion or the selector filter.

cc @flaviendelangle

**Review by flaviendelangle — COMMENTED** [review]
Nothing blocking for Claude Code
I'll test manually and look at the DX before approving 👍 

# PR review

This internal-only dependencies data model is in good shape: no correctness bugs survived a line-by-line pass, a bug/regression sweep, or verification of the trickiest paths (the controlled cascade, WeakMap slice caching, Map last-wins dedup, selector memoization, and the `SchedulerStore` seam). The head SHA `c9ce9fd` already folds in the self-dependency / duplicate-edge fixes from the referenced comment. Nothing is merge-blocking. What remains are a few test gaps and one API-naming nit worth settling before this surface flips public.

## Bugs (0)

No findings. I specifically confirmed as safe: the `handleEventsUpdate` cascade only fires from user-facing `updateEvents({ deleted })` (not lazy-loading eviction), so scrolling events out of range can't silently drop dependencies; `registerStoreEffect` fires only on subsequent changes, so the constructor's explicit `warnOnInvalidDependencies()` doesn't double-warn on mount; `.dataTimezone.rrule` can't null-deref (`dataTimezone` is required on `SchedulerProcessedEvent`); and `SchedulerAddDependencyResult` is a sound discriminated union (narrowing on `reason` correctly gates `eventId` vs `dependencyId`).

## Tests (2)

### 1. 🟡 Two-for-one `deleteDependency` under duplicate ids is unasserted

**Location:** `packages/x-scheduler-internals-premium/src/internals/plugins/SchedulerSchedulingPlugin.ts:135`

```ts
const remaining = current.filter((dependency) => dependency.id !== dependencyId);
this.updateDependenciesIfChanged(current, remaining);
```

Duplicate ids are tolerated in `dependencyModelList` (only the lookup dedups, last-wins — and that's covered). But `deleteDependency('dup-1')` filters by `id !==`, so when two entries share `dup-1` it removes **both** and emits a list two shorter. `updateDependenciesIfChanged` still fires (length dropped), so it "works," but the two-for-one semantics are never pinned.

**Failure scenario:** A future refactor changes `deleteDependency` to remove only the first match; nothing fails, and duplicate-id deletion silently regresses.

**Fix:** Add a test with `dependencies: [{id:'dup', a→b}, {id:'dup', b→a}]` asserting `deleteDependency('dup')` emits `[]`.

### 2. 🟡 In-render duplicate `addDependency` calls are undetectable — not pinned

**Location:** `packages/x-scheduler-internals-premium/src/internals/plugins/SchedulerSchedulingPlugin.ts:118`

The duplicate guard reads the controlled `this.store.state.dependencyModelList`, which is not written between calls (the existing "distinct id" test proves this). So `addDependency({a→c})` called twice in one render passes the guard both times and emits two `a→c` edges. The duplicate-rejection test only covers duplicating an edge already present in props.

**Failure scenario:** A consumer batches two identical `addDependency` calls before re-rendering and ends up with a duplicate edge — behavior no test documents as intended-or-not.

**Fix:** Add a test asserting the intended contract for back-to-back identical adds (and note it's a known controlled-prop limitation, same as events).

## Simplifications (2)

### 1. 🟡 `SchedulerDependencyRejectionReason` doesn't enumerate all rejection reasons

**Location:** `packages/x-scheduler-internals-premium/src/models/dependency.ts:40`

```ts
export type SchedulerDependencyRejectionReason = 'recurringEvent' | 'unknownEvent';

export type SchedulerAddDependencyResult =
  | { status: 'added'; id: SchedulerDependencyId }
  | { status: 'rejected'; reason: SchedulerDependencyRejectionReason; eventId: SchedulerEventId }
  | { status: 'rejected'; reason: 'duplicateDependency'; dependencyId: SchedulerDependencyId };
```

The `eventId`/`dependencyId` split is defensible (a duplicate rejection genuinely points at an existing edge, not an event). The nit is that a type literally named "rejection reason" omits `'duplicateDependency'`, so it overpromises: a helper typed `(reason: SchedulerDependencyRejectionReason)` can't be handed `result.reason` without a cast, and readers will assume it's exhaustive.

**Cost:** The set of "why rejected" values is fragmented across the enum and an inline literal; the next non-event reason repeats the inconsistency. Cheap to settle now, awkward to change once public.

**Fix:** Either widen the enum to include `'duplicateDependency'`, or rename it (e.g. to signal it's the *event-side* reasons) so the omission reads as intentional.

### 2. ℹ️ `type` excluded from duplicate identity — guarded only by a comment

**Location:** `packages/x-scheduler-internals-premium/src/internals/plugins/SchedulerSchedulingPlugin.ts:118`

Duplicate matching is `source === source && target === target`, ignoring `type` (the comment acknowledges it's fine while the union has one member). Once `SchedulerDependencyType` widens, `addDependency({a→b, StartToStart})` will be wrongly rejected as a duplicate of `a→b, FinishToStart`. Nothing but the comment prevents this from being forgotten. Worth a linked `// TODO(#…)` or an intent-pinning note so the guard grows with the union.

## Docs (0)

No findings — the JSDoc reword (`source`/`target` as graph direction, single-member `type` union with a widening note) matches the code, and the cascade persistence limitation is documented on `handleEventsUpdate`.

## Verdict

**Approve after nits** — no correctness or regression issues; only a couple of low-cost test gaps and one internal-API naming nit worth resolving before the surface goes public.

---

🤖 Review generated with Claude Code

**Comment by rita-codes** [comment]
Applied in `24d1f939f7`:

- **Rejection reasons:** restructured as `SchedulerDependencyEventRejectionReason` (`'recurringEvent' | 'unknownEvent'` — what the classifier returns and the `eventId` branch carries) with `SchedulerDependencyRejectionReason` as the exhaustive union including `'duplicateDependency'` — the name no longer overpromises and the classifier stays honest.
- **Duplicate identity:** the comment is now a `TODO(#22853)` so widening the type union picks it up.
- **Tests:** pinned the two-for-one `deleteDependency` under duplicate ids (the id is the identity, so deleting it removes all of its entries) and the same-tick duplicate-add bypass (same controlled-prop limitation as consecutive adds, now documented in the test).

cc @flaviendelangle

File: packages/x-scheduler-internals-premium/src/models/dependency.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * The other PDM types (`StartToStart`, `FinishToFinish`, `StartToFinish`) will widen this union when implemented.
+ */
+export type SchedulerDependencyType = 'FinishToStart';
+
+/**
+ * A dependency between two events, referencing them by id.
+ * For `"FinishToStart"`, `source` is the predecessor and `target` the successor.
+ */
+export interface SchedulerDependency {
```
Comment (flaviendelangle) [inline packages/x-scheduler-internals-premium/src/models/dependency.ts:15]: Will all PDM types avec the same properties here?

File: packages/x-scheduler-internals-premium/src/use-event-timeline-premium/tests/core.EventTimelinePremiumStore.test.ts [hunk 2]
```diff
@@ -22,6 +22,8 @@ describe('Core - EventTimelinePremiumStore', () => {
         canDragEventsFromTheOutside: false,
         canDropEventsToTheOutside: false,
         copiedEvent: null,
+        dependencyModelList: [],
+        dependencyModelLookup: new Map(),
```
Comment (flaviendelangle) [inline packages/x-scheduler-internals-premium/src/use-event-timeline-premium/tests/core.EventTimelinePremiumStore.test.ts:26]: Why store both the list and the lookup in state?
Couldn't we derive one from the other?

File: packages/x-scheduler-internals-premium/src/internals/plugins/SchedulerSchedulingPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    }
+  }
+
+  [disposeSymbol]() {
+    this.disposables.dispose();
+  }
+
+  private updateDependencies(newDependencies: SchedulerDependency[]) {
+    const eventDetails = createChangeEventDetails('none');
+    this.store.parameters.onDependenciesChange?.(newDependencies, eventDetails);
```
Comment (flaviendelangle) [inline packages/x-scheduler-internals-premium/src/internals/plugins/SchedulerSchedulingPlugin.ts:72]: Is it intended that the dependencies must be controlled?
Looks like we never update the state directly whe not controlled

File: packages/x-scheduler-internals-premium/src/models/dependency.ts (hunk 1 above)
Comment (rita-codes) [reply packages/x-scheduler-internals-premium/src/models/dependency.ts:15]: Yes! The four PDM types are all just a link between two events, so source/target/type works for all of them 🙌

File: packages/x-scheduler-internals-premium/src/internals/plugins/SchedulerSchedulingPlugin.ts (hunk 3 above)
Comment (rita-codes) [reply packages/x-scheduler-internals-premium/src/internals/plugins/SchedulerSchedulingPlugin.ts:72]: Intended, yes. Right now all data collections in the scheduler (like events) are controlled-only. Dependencies are domain data the app needs to persist, so same category IMO 👀

File: packages/x-scheduler-internals-premium/src/use-event-timeline-premium/tests/core.EventTimelinePremiumStore.test.ts (hunk 2 above)
Comment (rita-codes) [reply packages/x-scheduler-internals-premium/src/use-event-timeline-premium/tests/core.EventTimelinePremiumStore.test.ts:26]: A selector would work too! The main reason is that the Map gets built anyway while validating the list (it's how duplicate ids are detected, same as `buildEventsState` does for events), so storing it is free and a selector would redo the same work. And it keeps the shape consistent with events, where both live in state. Wdyt?

**Review by flaviendelangle — APPROVED** [review]
